src/functional/ffront/foast_passes/type_deduction.py

```diff
@@ -107,8 +260,177 @@ def visit_FieldSymbol(
+                op, parent=parent, left_type=left_type, right_type=right_type, **kwargs
+            )
+        else:
+            return self._deduce_logical_binop_type(
+                op, parent=parent, left_type=left_type, right_type=right_type, **kwargs
+            )
+
+    def _binop_to_str(self, op: foast.BinaryOperator) -> str:
+        match op:
+            case foast.BinaryOperator.ADD:
+                return "+"
+            case foast.BinaryOperator.SUB:
+                return "-"
+            case foast.BinaryOperator.MULT:
+                return "*"
+            case foast.BinaryOperator.DIV:
+                return "/"
+            case foast.BinaryOperator.BIT_AND:
+                return "&"
+            case foast.BinaryOperator.BIT_OR:
+                return "|"
+
+    def _deduce_arithmetic_binop_type(
+        self,
+        op: foast.BinaryOperator,
+        *,
+        parent: foast.BinOp,
+        left_type: foast.SymbolType,
+        right_type: foast.SymbolType,
+        **kwargs,
+    ) -> foast.SymbolType:
+        left, right = TypeInfo(left_type), TypeInfo(right_type)
+        if (
+            left.is_arithmetic_compatible
+            and right.is_arithmetic_compatible
+            and dims_broadcast_compatible(left, right)
+        ):
+            return dims_broadcasted(left, right).type
+        else:
+            raise FieldOperatorTypeDeductionError.from_foast_node(
+                parent,
+                msg=f"Incompatible type(s) for operator '{self._binop_to_str(op)}': {left.type}, {right.type}!",
+            )
+
+    def _deduce_logical_binop_type(
+        self,
+        op: foast.BinaryOperator,
+        *,
+        parent: foast.BinOp,
+        left_type: foast.SymbolType,
+        right_type: foast.SymbolType,
+        **kwargs,
+    ) -> foast.SymbolType:
+        left, right = TypeInfo(left_type), TypeInfo(right_type)
+        if (
+            left.is_logics_compatible
+            and right.is_logics_compatible
+            and dims_broadcast_compatible(left, right)
+        ):
+            return dims_broadcasted(left, right).type
+        else:
+            raise FieldOperatorTypeDeductionError.from_foast_node(
+                parent,
+                msg=f"Incompatible type(s) for operator '{self._binop_to_str(op)}': {left.type}, {right.type}!",
+            )
+
+    def visit_UnaryOp(self, node: foast.UnaryOp, **kwargs) -> foast.UnaryOp:
+        new_operand = self.visit(node.operand, **kwargs)
+        if not self._unaryop_type_compatible(op=node.op, operand_type=new_operand.type):
+            raise FieldOperatorTypeDeductionError.from_foast_node(
+                node,
+                msg=f"Incompatible type for unary operator '{self._unaryop_to_str(node.op)}': {new_operand.type}!",
+            )
+        return foast.UnaryOp(
+            op=node.op, operand=new_operand, location=node.location, type=new_operand.type
+        )
+
+    def _unaryop_type_compatible(
```

Comment:
A predicate-like name seems always clearer to me.
```suggestion
    def _is_unaryop_type_compatible(
```